### DISCUSSION TITLE
Can OpenLineage be used for explaining data lineage inside a relational database?

### DISCUSSION THREAD

**Discussion by danielfernandez** [OpenLineage/OpenLineage#3218] [General]
I am new to OpenLineage and, from reading docs and articles, my impression is that it is well-adapted for modelling transformations of data _in motion_, but I'm struggling to see how to use it to document the kind of static data transformations that are defined in relational database systems by means of **views** and other structures.

A relational database may define a series of **views** (persisted or not), both based on tables and also on other views, and both inside a single database or accross several databases, that can end up creating a large lineage tree of, potentially, dozens of levels.

So we are talking about the typical views created with `CREATE VIEW x AS SELECT ...` that are therefore defined by a SQL query that retrieves data from tables in the same RDBMS, or from other views also created with `CREATE VIEW x AS SELECT ...`, etc.

These views define _transformations_ of data, but they define those transformations in a _static_ manner, i.e. there are no _transformation events_ that could be modelled — unless we modelled as such an event every query performed on a view (not feasible).

The complexity and richness of such lineage trees is huge, and it can create very different intermediate and end-user datasets that serve very different profiles accross a company. It is very important for data stewards to understand those lineage trees and where the data in the final _products_ (end-user datasets) actually comes from.

This is a type of lineage that can be easily modelled in most Data Governance tools I've worked with but, as mentioned, I'm struggling to see how this can be adapted to the OpenLineage format. I see how RDBMS tables and/or views could be modelled as "datasets" using the `DatasetFacet`, but... what about the _static lineage relationship among them_?

The definition of views in a RDBMS and the data lineage relationships they create do not correspond to any specific _events_, nor to any movement of data from one system to another: it all happens _inside_ the RDBMS and all this lineage has a _static_ nature.

Any help would be appreciated.

**Comment by mobuchowski** [comment]
Hello @danielfernandez.

I think this is possible to model that using OpenLineage. The additional question is how to expose that - and I think most popular way amongst databases is to have specific system view that contains that information.

Addressing the main point:

>These views define transformations of data, but they define those transformations in a static manner, i.e. there are no transformation events that could be modelled — unless we modelled as such an event every query performed on a view (not feasible).

>The definition of views in a RDBMS and the data lineage relationships they create do not correspond to any specific events, nor to any movement of data from one system to another: it all happens inside the RDBMS and all this lineage has a static nature.

I feel that's not true: there is a discrete moment of time when view gets created: when the `CREATE VIEW` command gets executed.
This would be modeled as an regular event that contains input datasets: tables/views that are being used when the view gets created, and output dataset, which is the actual created view.

The addition that would enhance that model is standardized dataset facet that specify that this dataset is a non-materialized view, which would help the OpenLineage consumer understand that despite lack of direct updates, the underlying tables might be updated.

Then, subsequent events that read/write/transform using the view can use it as a regular input dataset without needing to present the whole lineage hierarchy, and it would a job of OpenLineage consumer to present information about that transparently to the user.

When the view gets updated, for example by adding new column, that would be an additional regular OpenLineage event that has output dataset representing that view. 

That being said, while OpenLineage tries to structure how the lineage gets represented, of course, each "data system" works differently and there are differences in how they would model their lineage if they implemented OpenLineage. One difference I could easily point to are temporary views created inside some more complicated transformation: as internals of particular job, that exist only within some transformation, they are not interesting or usable outside that job, so they should not be represented. 

Also, contributions are very welcome, so if you have an idea to make OpenLineage better in those situation, feel free to discuss it.

**Reply by danielfernandez** [reply]
Thanks for your comment.

> The additional question is how to expose that - and I think most popular way is to have specific system view that contains that information.

Could you specify what you mean with _"specific system view"_? You mean a specific way for visualization tools to represent this kind of lineage?

> I feel that's not true: there is a discrete moment of time when view gets created: when the CREATE VIEW command gets executed.
> This would be modeled as an regular event that contains input datasets: tables/views that are being used when the view gets created, and output dataset, which is the actual created view.

So modelling these relationships by means of Run State Updates using the Dataset Facets for column lineage, as if the _event_ being modelled was the creation of the view, right?

I thought about this as a possibility but, from the standpoint of someone consuming this OpenLineage data using a tool, wouldn't this end up being a bit confusing? Wouldn't it appear to users as a data transformation that happened _in the past_, as if someone had run an ETL job that resulted in data being actually moved from A to B, and never happened again?

The `eventTime` is a required property so, even if it were used for specifying the _creation time_ of the view... Is there any way to signal that this was not a one-time event, that this transformation stays in place and the represented lineage will affect any query performed on the _output_ dataset in the future, without needing to model each query as a separate event?

**Reply by mobuchowski** [reply]
>Could you specify what you mean with "specific system view"? You mean a specific way for visualization tools to represent this kind of lineage?

No, I mean a system view like ACCESS_HISTORY in Snowflake. Or, the database could emit the events themself. Or, [like Trino](https://trino.io/docs/current/develop/event-listener.html), add a mechanism that would allow users to write the custom integration themselves.

>I thought about this as a possibility but, from the standpoint of someone consuming this OpenLineage data using a tool, wouldn't this end up being a bit confusing? Wouldn't it appear to users as a data transformation that happened in the past, as if someone had run an ETL job that resulted in data being actually moved from A to B, and never happened again?

I think that's nice part about OpenLineage - there's nothing prescribing you how you would use or visualize that data. For example, [Astronomer use OpenLineage for alerting and SLAs.](https://www.astronomer.io/blog/introducing-astro-observe/) Your OpenLineage consumer could understand that this particular dataset is a view, and show "freshness" of it based on freshness of underlying tables, not the eventTime for this particular event. 

>The eventTime is a required property so, even if it were used for specifying the creation time of the view... Is there any way to signal that this was not a one-time event, that this transformation stays in place and the represented lineage will affect any query performed on the output dataset in the future, without needing to model each query as a separate event?

I think there are two questions here. First, there is no standard facet like that right now, but I think adding it is a good idea - would love to hear opinions from other people about that too.

I described how I think it would look like in the previous message:

> The addition that would enhance that model is standardized dataset facet that specify that this dataset is a non-materialized view, which would help the OpenLineage consumer understand that despite lack of direct updates, the underlying tables might be updated.

Second, I believe each query from end users would need to be modeled as an event - that reads those particular views - otherwise the lineage information wouldn't be exhaustive; you'd miss the actual usage pattern. If you wish to just model that view structure it's possible. 

Many databases already keep metadata about each query, like BigQuery, Snowflake, or like Trino allow plugins to act on any query - I don't think it's unreasonable unless you're thinking of high-performance OLTP systems, where I'm not sure the overall view structure fits.

**Comment by pawel-big-lebowski** [comment]
What it looks to me, the possible approaches are:
(1) view is represented as an output dataset of special kind job which is never run and consumer should be able to understand this (@mobuchowski comments above)
(2) views are resolved at the moment of creating OL event and inputs point into view's input tables instead of view definition

The problem with (1) is that if you don't emit static lineage events representing the views, you never know the view definition, assuming you created the view before turning OpenLineage on, which I expect to be the case most of the time. I think OL spec should be capable to contain view definition using a single OL event (like ViewDatasetFacet containing full view definition).

**Reply by danielfernandez** [reply]
Thanks for your answer.

The base issue for me is that, whatever the solution, I cannot make the system emit OL events for each query that reaches the system's views. i.e I cannot represent the _internal flow of data_ from original tables to the views appearing in the queries as lineage events. Or better said, it would not be _practical_ to do so.

The reason is twofold: 1. We are talking about systems getting several hundred queries per minute in production, and having to output full lineage information for each query would affect performance; and 2. Lineage trees for these views may have dozens of levels and therefore be super-complex, making the performance issue worse.

**Comment by Sheeri** [comment]
@danielfernandez A view does not have to be part of operational lineage (lineage based on what happened), it can be part of static lineage (aka "Design lineage", how the system is designed). I would use a DatasetEvent.

For static lineage - lineage that does not come from an operational "run" - there are 2 scenarios:
1) When you can associate a job with the lineage, use [JobEvent](https://github.com/OpenLineage/OpenLineage/blob/c4525608d67eb262a0b25909a627f45243666fa1/spec/OpenLineage.json#L80-L108).
2) When there is no associated job (e.g. a view), use [DatasetEvent](https://github.com/OpenLineage/OpenLineage/blob/c4525608d67eb262a0b25909a627f45243666fa1/spec/OpenLineage.json#L65-L79).

JobEvent and DatasetEvent were added specifically to accommodate static lineage. See the accepted proposal at https://github.com/OpenLineage/OpenLineage/blob/main/proposals/1837/static_lineage.md

ColumnLineage can be used inside of a DatasetEvent, and there's no status needed (e.g. "COMPLETE", "FAILED" etc)
